File: platforms/folia/src/main/java/com/cjcrafter/foliascheduler/folia/FoliaServer.java
package com.cjcrafter.foliascheduler.folia;

import com.cjcrafter.foliascheduler.AsyncSchedulerImplementation;
import com.cjcrafter.foliascheduler.EntitySchedulerImplementation;
import com.cjcrafter.foliascheduler.GlobalSchedulerImplementation;
import com.cjcrafter.foliascheduler.RegionSchedulerImplementation;
import com.cjcrafter.foliascheduler.ServerImplementation;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.entity.Entity;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

@ApiStatus.Internal
public class FoliaServer implements ServerImplementation {

    private final @NotNull Plugin owningPlugin;
    private final @NotNull FoliaGlobalScheduler globalScheduler;
    private final @NotNull FoliaAsyncScheduler asyncScheduler;

    @ApiStatus.Internal
    public FoliaServer(@NotNull Plugin owningPlugin) {
        this.owningPlugin = owningPlugin;
        this.globalScheduler = new FoliaGlobalScheduler(owningPlugin);
        this.asyncScheduler = new FoliaAsyncScheduler(owningPlugin);
    }

    @Override
    public @NotNull Plugin getOwningPlugin() {
        return owningPlugin;
    }

    @Override
    public boolean isOwnedByCurrentRegion(@NotNull Location location) {
        return Bukkit.isOwnedByCurrentRegion(location);
    }

    @Override
    public boolean isOwnedByCurrentRegion(@NotNull Location location, int squareRadiusChunks) {
        return Bukkit.isOwnedByCurrentRegion(location, squareRadiusChunks);
    }

    @Override
    public boolean isOwnedByCurrentRegion(@NotNull Block block) {
        return Bukkit.isOwnedByCurrentRegion(block);
    }

    @Override
    public boolean isOwnedByCurrentRegion(@NotNull World world, int chunkX, int chunkZ) {
        return Bukkit.isOwnedByCurrentRegion(world, chunkX, chunkZ);
    }

    @Override
    public boolean isOwnedByCurrentRegion(@NotNull World world, int chunkX, int chunkZ, int squareRadiusChunks) {
        return Bukkit.isOwnedByCurrentRegion(world, chunkX, chunkZ, squareRadiusChunks);
    }

    @Override
    public boolean isOwnedByCurrentRegion(@NotNull Entity entity) {
        return Bukkit.isOwnedByCurrentRegion(entity);
    }

    @Override
    public @NotNull GlobalSchedulerImplementation global() {
        return globalScheduler;
    }

    @Override
    public @NotNull AsyncSchedulerImplementation async() {
        return asyncScheduler;
    }

    @Override
    public @NotNull EntitySchedulerImplementation entity(@NotNull Entity entity) {
        return new FoliaEntityScheduler(owningPlugin, entity);
    }

    @Override
    public @NotNull RegionSchedulerImplementation region(@NotNull World world, int chunkX, int chunkZ) {
        return new FoliaRegionScheduler(owningPlugin, world, chunkX, chunkZ);
    }

    @Override
    public void cancelTasks() {
        globalScheduler.cancelTasks();
        asyncScheduler.cancelTasks();
    }
}

File: platforms/folia/src/main/kotlin/com/cjcrafter/scheduler/folia/FoliaServer.kt
package com.cjcrafter.scheduler.folia

import com.cjcrafter.scheduler.AsyncSchedulerImplementation
import com.cjcrafter.scheduler.EntitySchedulerImplementation
import com.cjcrafter.scheduler.ServerImplementation
import com.cjcrafter.scheduler.SchedulerImplementation
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.entity.Entity
import org.bukkit.plugin.Plugin

/**
 * The Folia provider implementation for multithreaded scheduling.
 */
class FoliaServer(override val owningPlugin: Plugin) : ServerImplementation {
    override fun isOwnedByCurrentRegion(location: Location): Boolean {
        return owningPlugin.server.isOwnedByCurrentRegion(location)
    }

    override fun isOwnedByCurrentRegion(location: Location, squareRadiusChunks: Int): Boolean {
        return owningPlugin.server.isOwnedByCurrentRegion(location, squareRadiusChunks)
    }

    override fun isOwnedByCurrentRegion(block: Block): Boolean {
        return owningPlugin.server.isOwnedByCurrentRegion(block)
    }

    override fun isOwnedByCurrentRegion(world: World, chunkX: Int, chunkZ: Int): Boolean {
        return owningPlugin.server.isOwnedByCurrentRegion(world, chunkX, chunkZ)
    }

    override fun isOwnedByCurrentRegion(world: World, chunkX: Int, chunkZ: Int, squareRadiusChunks: Int): Boolean {
        return owningPlugin.server.isOwnedByCurrentRegion(world, chunkX, chunkZ, squareRadiusChunks)
    }

    override fun isOwnedByCurrentRegion(entity: Entity): Boolean {
        return owningPlugin.server.isOwnedByCurrentRegion(entity)
    }

    override fun global(): SchedulerImplementation {
        return FoliaGlobalScheduler(owningPlugin)
    }

    override fun async(): AsyncSchedulerImplementation {
        return FoliaAsyncScheduler(owningPlugin)
    }

    override fun entity(entity: Entity): EntitySchedulerImplementation {
        return FoliaEntityScheduler(owningPlugin, entity)
    }

    override fun region(world: World, chunkX: Int, chunkZ: Int): SchedulerImplementation {
        return FoliaRegionScheduler(owningPlugin, world, chunkX, chunkZ)
    }
}
